File: app/src/main/java/cn/xhuww/mvvm/widget/CountDownView.java
package cn.xhuww.mvvm.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.support.v7.widget.AppCompatTextView;

/**
 * Created by xhu_ww on 2018/5/9.
 * description:倒计时View
 */
public class CountDownView extends AppCompatTextView implements Runnable {
    private boolean run = false;
    private String endHint;
    private int count;
    private boolean isCanRun;

    public CountDownView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CountDownView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public CountDownView(Context context) {
        super(context);
    }

    /**
     * 设置倒计完成时的显示提示
     *
     * @param endHint
     */
    public void setEndHint(String endHint) {
        this.endHint = endHint;
    }

    public boolean isRun() {
        return run;
    }

    public void setCount(int count) {
        this.count = count;
    }

    public void stopRun() {
        this.run = false;
    }

    public void startRun() {
        this.run = true;
        run();
    }

    @Override
    public void run() {
        if (run && count > 0) {
            this.setText(count + "s");
            postDelayed(this, 1000);
            count--;
        } else {
            run = false;
            if (endHint != null) {
                this.setText(endHint);
            } else {
                this.setText("0s");
            }
        }
    }
}

File: kotlin_mvvm/src/main/java/com/example/study/widget/CountDownView.kt
package com.example.study.widget

import android.annotation.SuppressLint
import android.content.Context
import android.support.v7.widget.AppCompatTextView
import android.util.AttributeSet

/**
 * Created by xhu_ww on 2018/6/11.
 * description: 倒计时控件
 */
class CountDownView @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : AppCompatTextView(context, attrs, defStyleAttr), Runnable {

    var run = false
    var endHint: String = ""
    var count: Int = 0

    @SuppressLint("SetTextI18n")
    override fun run() {
        if (run && count > 0) {
            this.text = count.toString() + "s"
            postDelayed(this, 1000)
            count--
        } else {
            run = false
            if (endHint.isEmpty()) {
                this.text = endHint
            } else {
                this.text = "0s"
            }
        }
    }

    fun stopRun() {
        this.run = false
    }

    fun startRun() {
        this.run = true
        run()
    }
}